mono/metadata/mono-perfcounters.c

```diff
@@ -1483,231 +1572,292 @@ typedef struct {
 	int type;
 } CounterCreationData;
 
+TYPED_HANDLE_DECL (CounterCreationData);
+
+static char*
+append (char* cursor, const char* str)
+{
+	int const length = strlen (str) + 1;
+	return memcpy (cursor, str, length) + length;
+}
+
 /*
  * Since we'll keep a copy of the category per-process, we should also make sure
  * categories with the same name are compatible.
  */
 MonoBoolean
-mono_perfcounter_create (MonoString *category, MonoString *help, int type, MonoArray *items)
+ves_icall_System_Diagnostics_PerformanceCounterCategory_Create (
+	MonoStringHandle category, MonoStringHandle help, int type, MonoArrayHandle items)
 {
+	HANDLE_FUNCTION_ENTER ()
+
 	MonoError error;
+	error_init (&error);
+
 	int result = FALSE;
-	int i, size;
-	int num_counters = mono_array_length (items);
-	int counters_data_size;
+	int const num_counters = mono_array_handle_length (items);
 	char *name = NULL;
 	char *chelp = NULL;
 	char **counter_info = NULL;
-	char *p;
-	SharedCategory *cat;
+	gboolean unlock = FALSE;
 
 	/* FIXME: ensure there isn't a category created already */
-	name = mono_string_to_utf8_checked (category, &error);
-	if (!mono_error_ok (&error))
-		goto failure;
-	chelp = mono_string_to_utf8_checked (help, &error);
-	if (!mono_error_ok (&error))
-		goto failure;
+	name = mono_string_handle_to_utf8 (category, &error);
+	goto_if_nok (&error, failure);
+
+	chelp = mono_string_handle_to_utf8 (help, &error);
+	goto_if_nok (&error, failure);
+
 	counter_info = g_new0 (char*, num_counters * 2);
 	/* calculate the size we need structure size + name/help + 2 0 string terminators */
+	int size;
 	size = G_STRUCT_OFFSET (SharedCategory, name) + strlen (name) + strlen (chelp) + 2;
+
+	int i;
 	for (i = 0; i < num_counters; ++i) {
-		CounterCreationData *data = mono_array_get (items, CounterCreationData*, i);
-		counter_info [i * 2] = mono_string_to_utf8_checked (data->name, &error);
-		if (!mono_error_ok (&error))
-			goto failure;
-		counter_info [i * 2 + 1] = mono_string_to_utf8_checked (data->help, &error);
-		if (!mono_error_ok (&error))
-			goto failure;
+		MONO_HANDLE_LOCAL_VARIABLE_INITIALIZED_NULL (CounterCreationData, data);
```

Comment:
See comment below about loop bodies that allocate handles.

(I missed this loop first time through, so pardon the out of order comments)